docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+
+### Goals
+
+- Define an API contract that
+    - allows infrastructure providers to dynamically allocate and release addresses
+    - allows to write IPAM providers to integrate with any IPAM solution
+    - supports both IPv4 and IPv6
+
+### Non-Goals/Future Work
+
+- IPAM providers should not be added to CAPI directly, but live as individual projects instead.
+- IPAddress and IPAddressPool will only have limited validation (e.g. that the referenced Pool exists and that IP address strings are valid addresses). Advanced validation (whether an IP is actually part of the correct subnet for example) needs to be handled by IPAM providers.
+
+## Proposal
+
+The chosen approach was heavily inspired by metal3's ip-address-manager, which in turn seems to have been inspired by Kubernetes' PersistentVolumes. The proposal is twofold and consists of the API contract that should be added to CAPI and two examples to better illustrate how the contract is used by IPAM and infrastructure providers.
+
+### IPAM API Contract
+
+The IPAM API contract provides a generic interface between CAPI infrastructure providers and to-be-developed IPAM providers. Any intersted infrastructure is supposed to be able to request IP Addresses from a defined pool of addresses. IPAM providers should then be able to find and fulfil such requests using any custom logic to do so.
+
+The contract consists of three resources:
+
+- An IPAddressClaim, which is used to allocate addresses
+- An IPAddress that IPAM providers use to fulfil IPAddressClaims
+- IP Pools, which hold provider specific configuration on how addresses should be allocated, e.g. a specific subnet to use
+
+Both **IPAddressClaims** and **IPAddresses** should be part of Cluster API, while **IPPools** are defined by the different IPAM providers.
+
+An **IPAddressClaim** is used by infrastructure providers to request an IP Address. The claim contains a reference to an IP Pool. Because the pool is provider specific, the IPAM controllers can decide whether to handle a claim by inspecting the group and kind of the pool reference.
+
+If a IPAM controller detects a Claim that references a Pool it controls, it allocates an IP address from that pool and creates an **IPAddress** to fulfil the claim. It also updates the status of the IPAddressClaim with a reference to the created Address.
+
+```go
+// IPAddressClaimSpec describes the desired state of an IPAddressClaim
+type IPAddressClaimSpec struct {
+   // Pool is a reference to the pool from which an IP address should be allocated.
```

Comment:
Given an IP Pool is an arbitrary API of choice for the IPAM providers, wouldn't it be possible for multiple IPAM providers to coexist and have pools with the same name? i.e wouldn't this need to be a TypedLocalObjectReference?

Reply:
Good point, this is not really clear here: From the work on moving away from v1.ObjectReference I had some opinion on those reference types. This one would actually consist of Kind, Name and APIVersion. Not sure if that makes sense.

I planned to open a PR that adds those types to the current v1beta1 API to be able to be used for further additions to that version, as well as to hopefully migrate to them with v1beta2 (even though its arguably a breaking change). I think with the types already being there, and discussed already, it would be a lot easier to migrate, since the discussion doesn't have to be part of it anymore.

Reply:
I've changed it to Group, Kind and Name. Version is explicitly excluded as it actually does not make sense to have that. The client should decide on it's own which version it wants to consume.